### PR TITLE
Test code coverage publishing

### PR DESCRIPTION
**Pull request by srnagar**
This PR enables publishing code coverage report to DevOps build pipeline tab. The intent of this change is to collect aggregate test coverage report for track 2 client libraries under each service directory (sdk/<servicearea>). In order to generate this as an aggregate report at the service level, this PR makes 2 main changes:

## Update to service pom.xml
Since JaCoCo's aggregate report [does not work on an aggregator module](https://github.com/jacoco/jacoco/wiki/MavenMultiModule#strategy-aggregator) (i.e. the existing service pom.xml), the official recommended strategy is to create a dummy module to include all track 2 modules of a service as a dependency and use this dummy module to generate the aggregate report. However, I was able to repurpose the aggregator `pom.xml` by using profiles to make the aggregator pom look like the dummy module when the `coverage` profile is activated by including `-Pcoverage` option when running the maven command. By default, the service pom.xml will just be an aggregator, as it previously was, which includes only a list of modules. When generating the report, the `coverage` profile will be activated which will then have the dependencies and the maven goal to generate the jacoco report.

## Update the client pipeline yml to publish coverage report
To publish the coverage report, there are 2 new steps added to the pipeline
- Generate the aggregate report: This step is run immediately after the unit tests are run. This step uses the service pom.xml and activates the `coverage` profile and targets the `jacoco:report-aggregate` maven goal to generate the `jacoco.xml` file in sdk/<service>/target/site/jacoco-coverage` directory.
- Publish the report - this is a standard DevOps step that supports publishing JaCoCo reports.

Both these steps are enabled to run only for `Windows` + `JDK 11` combination.

This PR enables publishing test coverage for `template` as proof of concept. If this is an acceptable solution, I'll extend it to other services.

fixes https://github.com/Azure/azure-sdk-for-java/issues/17316
fixes https://github.com/Azure/azure-sdk-for-java/issues/17317

### CODE REVIEW

File: sdk/eventhubs/pom.xml [hunk 1]
```diff
@@ -7,11 +7,68 @@
...
+    </profile>
+    <profile>
+      <id>default</id>
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <modules>
+        <module>microsoft-azure-eventhubs</module>
+        <module>microsoft-azure-eventhubs-eph</module>
+        <module>microsoft-azure-eventhubs-extensions</module>
```
Comment (alzimmermsft) [inline sdk/eventhubs/pom.xml:68]: Should we think of having a `pom.data.xml`, in a similar vain as we have `pom.mgmt.xml` to separate out Track 1 libraries from Track 2?

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 2]
```diff
@@ -306,6 +306,20 @@ jobs:
         # we want to run this when TestFromSource isn't true
         condition: and(succeeded(), ne(variables['TestFromSource'],'true'))
 
+      - task: Maven@3
+        displayName: 'Generate aggregate code coverage report'
+        condition: and(eq(variables['OsName'], 'Windows'), eq(variables['JavaTestVersion'], '1.11'))
```
Comment (alzimmermsft) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:311]: Do we only want to run this on Windows or Linux? Linux would run much faster, if that is a concern.

File: sdk/core/pom.xml [hunk 3]
```diff
@@ -8,19 +8,137 @@
...
-    <module>azure-core-serializer-avro-apache</module>
-    <module>azure-core-serializer-json-gson</module>
-    <module>azure-core-serializer-json-jackson</module>
-    <module>azure-core-test</module>
-    <module>azure-core-tracing-opentelemetry</module>
-  </modules>
+
+  <profiles>
+    <!-- DO NOT CHANGE THE *ORDER* IN WHICH THESE PROFILES ARE DEFINED! -->
+    <profile>
```
[the remark this answers is not in the document]
Comment (srnagar) [reply sdk/core/pom.xml:14]: We also want to restrict the modules to track 2 modules and not include track 1 stuff.

File: sdk/core/pom.xml [hunk 4]
```diff
@@ -8,19 +8,137 @@
...
+                    <exclude>**/com/azure/cosmos/implementation/apachecommons/**/*</exclude>
+                    <exclude>**/com/azure/cosmos/implementation/guava25/**/*</exclude>
+                    <exclude>**/com/azure/cosmos/implementation/guava27/**/*</exclude>
+                  </excludes>
+                </configuration>
+              </execution>
+            </executions>
+          </plugin>
+        </plugins>
+      </build>
```
[the remark this answers is not in the document]
Comment (srnagar) [reply sdk/core/pom.xml:120]: Having the parent pom would be my preference too but will hold off until some from EngSys confirms that they are okay with adding another parent pom for service level `pom.xml`.

File: sdk/eventhubs/pom.xml (hunk 1 above)
Comment (srnagar) [reply sdk/eventhubs/pom.xml:68]: That's another question for EngSys.

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml (hunk 2 above)
Comment (srnagar) [reply eng/pipelines/templates/jobs/archetype-sdk-client.yml:311]: Well, only one job can publish the report. If multiple jobs publish, the last one wins and invariably the "Build from source" job that runs on Linux and JDK 11 will be the last and publishes empty results. Given that this is only collecting report from already created jacoco reports from individual modules, speed is not a huge concern and Windows or Linux doesn't make a huge difference.

File: sdk/core/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (mitchdenny) [reply sdk/core/pom.xml:120]: Adding a parent pom for core probably doesn't add a lot of value. If it wasn't something that we were going to apply broadly to other "service areas" as well I definitely wouldn't want to do it. It just makes automation a hodgepodge of special cases.

**Comment by mitchdenny** [comment]
Will running code coverage after the unit tests are run effectively result in another test run? That seems kind of wasteful, can we just add this option to the unit test run?

**Comment by mitchdenny** [comment]
Had a chat with @srnagar. The rough plan of action is that we:

1. Go with the profile model in the service aggregator POM.
2. Experiment first with ```sdk/template/pom.xml``` to flush out any issues in the release side of the pipeline.
3. Let it settle for a bit just to make sure no issues crop up.
4. Come back and apply the same technique to core.

File: sdk/core/pom.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (srnagar) [reply sdk/core/pom.xml]: Discussed this with @mitchdenny offline and we decided not to have a parent pom for service areas as that would require shipping the parent pom to maven. So, we just have to duplicate the jacoco config.

File: sdk/eventhubs/pom.xml (hunk 1 above)
Comment (srnagar) [reply sdk/eventhubs/pom.xml]: This may be out of scope for this PR.

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**